### PR TITLE
Implement CuDNN Attention Backend with Graph Caching

### PR DESCRIPTION
**Pull request by ykdcai**
## Motivation

We implemet a CuDNN based backend for SGLang. Comparing to the previous attempt #2272 , this time we create the CuDNN pygraphs in AttentionBackend creation, reducing the overhead of graph creation in extend and decode significantly. Benchmark results show the throughput CuDNN backend is around 10% lower than flashinfer, maybe due to padding.

## Modifications

We create a `CuDNNBackend`. In the `__init__` method, the CuDNN execution graphs are created:
1. `self._decode_graphs`: Given the `max_batch_size`, create `max_batch_size/step_len` decode graphs.
2. `self._prefill_graphs`: Given `max_sequence_length`, create `(max_squence_length/step_len)^2` graphs: Note that **quadratic** number of graphs is required because there are two parameter that changes: length of query and `diagonal_band_right_bound` (The starting offset of the causal mask).

In `forward_extend`, we use for loop, iterating over the batch dimension and computing attention for each sequence. This might seem less efficient. However, for-loop seems necessary for correct *causality*: the `extend_prefix_len` of each seqeunce is different, yet `diagonal_band_right_bound` argument is just one integer that will be applied to all sequences in the batch. Alternatively, we can pad in front of each sequence, but it will increase memory overhead significently if `extend_prefix_len` of the sequences within the batch varies a lot. In the future, if CuDNN support using an array as `diagonal_band_right_bound`, we can remove the for loop. In `forward_decode`, we use a single `graph.execute()` to decode tokens for all the sequences. 

Aside from this, CuDNN graph constuction time can be quite long: one CuDNN graph costs around 0.4s. What's more, since extend needs to cache `O(n^2)` graphs, with max sequence length 1024 and step length 64, it takes around 100s. We are now trying using **Python multiprocessing** to accelerate graph creation.

We also include another `cudnn_backend_test.py` file with a mocked backend for easier debugging and profiling. This file will be removed when the PR is ready to merge.

## Benchmark Results
We run `python3 -m sglang.bench_one_batch --attention-backend cudnn --batc
h-size 32 --input 1024 --output 128 --model-path meta-llama/Llama-3.1-8B-Instruct --disable-cuda-graph ` on one H100-80G GPU. The benchmark results are shown below:
| Item              | Prefill Latency(s) | Prefill Throughput(token/s) | Decode Latency(s) | Decode Throughput (token/s) |
| :------------------------ | :------: | :----: | :----: | :----: |
| FlashInfer |   0.725 | 45137 |  0.009   | 3205 |
| CuDNN |   0.870 | 37563 | 0.011   | 2817 |
| CuDNN (CUDA Graph Decode) |   0.870 | 37563 | 0.009   | 3720 |
| CuDNN (No Loop) | 0.801 | 40139 | 0.011  | 2817  | 

The performance of CuDNN is roughly 15% slower than FlashInfer for extend: this result is only partial due to using loop: as shown on this table, we also have a loop-free baseline using ragged tensors. This baseline uses ragged tensor yet only supports `causal=False`, making it unfit for prefill. Nevertheless, the loop free version is only marginally faster than for-loop version. This shows the overhead of padding also plays a role. Aside from this, we will also try launching the kernels in different cuda streams to improve parallelism in the for loops.

## Checklist

- [X ] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [X ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [X] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [X] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

### CODE REVIEW

**Comment by yhyang201** [comment]
I previously attempted to integrate the CuDNN backend for SGLang, so I might have some experience to share. 

1. The following configuration:
```python
            diagonal_band_right_bound=0 if is_causal else None,
            diagonal_alignment=(
                cudnn.diagonal_alignment.BOTTOM_RIGHT
                if is_causal
                else cudnn.diagonal_alignment.TOP_LEFT
            ),
```
may help avoid the need to manually adjust the starting offset of the causal mask.


2. In fact, if the configuration in point 1 is used, does that mean it's possible to support `causal=True` when using ragged tensors? In my previous implementation, I didn’t find any constraints suggesting that ragged tensors only support `causal=False`.

3. In my earlier experiments, the batch size (`bs`) dimension appeared to impact performance. That is, using a graph built with `bs=64` for an actual `bs=1` extension didn’t yield significantly better performance compared to using a graph built directly with `bs=1`. (Of course, you could run your own experiments to verify whether this observation holds true.)

Because of this, I had considered building separate graphs for both `bs` and `seq`, which would still result in a quadratic number of graphs. Given the overhead of graph construction and the relatively modest performance gains, I decided not to pursue this direction further.

As for the idea of using multi-threading, I’d recommend checking with the repository maintainers to confirm whether that would be acceptable.

This issue https://github.com/sgl-project/sglang/issues/5064 might suggest that there are still ongoing efforts toward CuDNN integration for Blackwell.

As this integration challenge has been something I’ve struggled with for quite some time, I’m genuinely glad to see that others are still actively working on it and coming up with better solutions. I truly hope you're able to make progress — and if there’s anything I can do to help, please don’t hesitate to reach out.

**Comment by ykdcai** [comment]
Thank you for your suggestion! We did reference your implementation for our CuDNN backend.
1. Regarding the first point on `diagonal_band_right_bound`, it seems our current implementation is required to pass the correctness test: when calculating attention, we want the `i`th token in query to attend to the first `i+extend_prefil_len[i]` tokens for causality, which means `diagonal_band_right_bound` should be set to `extend_prefix_len[i]`. This also explains why we need a for loop in forward extend: since `extend_prefix_len` value are different for queries in the same batch and the `diagonal_band_right_bound` can only be an integer, we need seperate graphs for each sequence. The same problem is also true for ragged tensor. *If CuDNN could support using a tensor of size `batch_size` as `diagonal_band_right_bound` rather than a single integer, the for loop would not be needed*.
2. For the performance impact on batch size, we will keep that in mind. We are running some profiling and will update our implementation based on profiling results.

File: python/sglang/test/cudnn_backend_benchmark.py [hunk 1]
```diff
@@ -0,0 +1,473 @@
...
+import time
+from dataclasses import dataclass
+
+import cudnn
+import numpy as np
+import psutil
+import torch
+
+# 添加项目根目录到Python路径
+project_root = os.path.dirname(
```
Comment (yyihuang) [inline python/sglang/test/cudnn_backend_benchmark.py]: update comments to English

**Comment by Fridge003** [comment]
Do you have plan of supporting Cuda Graph for CuDNN backend? It's crucial for performance.

**Comment by ykdcai** [comment]
> Do you have plan of supporting Cuda Graph for CuDNN backend? It's crucial for performance

Yes, we are currently working on CUDA Graph support.

**Comment by ykdcai** [comment]
Just added cuda graph support for decode. Now CuDNN backend (3720 tokens/s) can outperform flashinfer (3205 tokens/s) in decode.

**Comment by DevashishLal-CB** [comment]
Looking forward to this getting integrated

**Comment by ykdcai** [comment]
The backend suffer from one correctness issue in the decoding phase:  as shown below, the output at even position of a batch are gibberish while the output at odd position is fine. The problem can be replicated by running ` SGLANG_ENABLE_TORCH_COMPILE=0 python3 -m sglang.bench_one_batch --attention-backend cudnn --batch-size 32 --input 1024 --output 128 --model-path meta-llama/Llama-3.1-8B-Instruct --disable-cuda-graph --correctness-test --page-size 1` 
```
========== Prompt 0 ==========
<|begin_of_text|>The capital of France is a city of romance, art, fashion, and cuisine. Paris is a must-visit destination for anyone who loves history, architecture, and culture. From the iconic Eiffel Tower to the world-famous Louvre Museum, Paris has something to offer for every interest and age.
The city is divided into 20 arrondissements, each with its own unique character and charm. The Latin Quarter is a popular area for students and young travelers, while the Champs-Élysées is a hub for shopping and dining. The Montmartre neighborhood is famous for its bohemian vibe and stunning views of the city.
Paris 

========== Prompt 1 ==========
<|begin_of_text|>The capital of the United Kindom is London, thefirst, thefirst, thefirst, thefirst, thefirst, thefirst, thefirst, thefirst, thefirst, thefirst, thefirst, thefirst, thefirst, thefirst, thefirst, thefirst, thefirst, thefirst, thefirst, thefirst, thefirst, thefirst, thefirst, thefirst, thefirst, thefirst, thefirst, thefirst, thefirst, thefirst, thefirst, thefirst, thefirst, thefirst, thefirst, thefirst, thefirst, thefirst, thefirst, thefirst, thefirst, thefirst, 

========== Prompt 2 ==========
<|begin_of_text|>Today is a sunny day and I like to take a walk in the park. I put on my sunglasses and my favorite hat and I'm ready to go. I take a deep breath and I feel the fresh air fill my lungs. I love days like this.
As I walk, I notice the beautiful flowers and trees around me. The birds are singing and the bees are buzzing from flower to flower. It's so peaceful and serene. I feel my worries and stress melting away with each step.
I take a seat on a bench and watch the world go by. I see children playing and laughing with their parents. I see couples holding hands and strolling together. I see 

========== Prompt 3 ==========
<|begin_of_text|>The capital of China is Beijing, thefirst, thefirst, thefirst, thefirst, thefirst, thefirst, thefirst, thefirst, thefirst, thefirst, thefirst, thefirst, thefirst, thefirst, thefirst, thefirst, thefirst, thefirst, thefirst, thefirst, thefirst, thefirst, thefirst, thefirst, thefirst, thefirst, thefirst, thefirst, thefirst, thefirst, thefirst, thefirst, thefirst, thefirst, thefirst, thefirst, thefirst, thefirst, thefirst, thefirst, thefirst, thefirst, 

```

The problem is very likely caused by some problem in the **decoding stage** (`forward_decode`), because if we use CuDNN for prefill and another backend for decode (like `torch_native`), the output is correct. You can verify by running:
```
SGLANG_ENABLE_TORCH_COMPILE=0 python3 -m sglang.bench_one_batch --decode-attention-backend torch_native
 --prefill-attention-backend cudnn --batch-size 32 --input 1024 --output 128 
--model-path meta-llama/Llama-3.1-8B-Instruct --disable-cuda-graph --correctness-test
```

However, since the output at odd position of the batch are correct, it is unlikely the the bug is caused by the incorrect computation logic inside `forward_decode`. Instead, it might be caused by errors in handling kv cache read or write. Note that **CuDNN runs in a seperate CUDA stream by default**, so there might be *race condition in kv cache handling or in the scheduler*, where multi cuda streams might be used (for instance, the `set_kv_cache` method uses two cuda streams). I tried binding CuDNN hande to the default PyTorch stream but it didn't fix the problem.

Solving this bug would require insight into kv cache handling and the schedule so I would appreciate it if somebody familiar with those code could took a look at this issue.

**Comment by ykdcai** [comment]
The correctness problem is solved thanks to @b8zhong: the sequence lengths expeceted by CuDNN should be `int32` yet sglang uses `int64`. The `seq_len_kv` must be converted to `int32` before feeding into CuDNN. With this bug fixed, the PR should be finished soon after I add support for CUDA graph
